[dtensor][8/N] Introduce cost model for sharding

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #111278
* #111234
* #110717
* __->__ #109145

This PR adds some basic comm cost model for sharding prop

wanchaol committed:
[dtensor][8/N] Introduce cost model for sharding

This PR adds some basic comm cost model for sharding prop

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][8/N] Introduce cost model for sharding"


This PR adds some basic comm cost model for sharding prop

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][8/N] Introduce cost model for sharding"


This PR adds some basic comm cost model for sharding prop

[ghstack-poisoned]

Chillee commented on torch/distributed/_tensor/_collective_utils.py:
```diff
@@ -158,3 +160,139 @@ def mesh_all_to_all(
+
+def spec_to_bytes(spec: "placement_types.DTensorSpec") -> int:
+    assert spec.tensor_meta is not None, "spec should have tensor meta defined!"
+    dtype_to_bytes = {
```
You can get `tensor.dtype.itemsize`

wanchaol replied:
cool good to know there's already a API for it, I'll update

wanchaol commented on torch/distributed/_tensor/_collective_utils.py:
```diff
@@ -158,3 +160,139 @@ def mesh_all_to_all(
+        / num_devices_on_mesh_dim
+    )
+
+
```
Here the costs are suitable for both intra node or inter-node or mixed, this is coming from the device mesh layout information where we know whether we can in the intra or inter node scenario

wanchaol committed:
Update on "[dtensor][8/N] Introduce cost model for sharding"


This PR adds some basic comm cost model for sharding prop. Why we need this? operators can generate multiple placement strategies, i.e. for matmul we have at least 4 possible shardings: `1. Shard(0), R 2. R, Shard(1) 3. Shard(1), Shard(0) 4. R, R`

We need to be able to choose from one of these options during runtime, and perform resharding with reasonable choices. This is why we are building a cost model for sharding here. In this PR we associate each possible sharding strategy with redistribute costs. For eager mode since we run ops eagerly, we simply perform a min cost selection. One can imagine if we have some global information the strategy selection would become more intelligient.


[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][8/N] Introduce cost model for sharding"


This PR adds some basic comm cost model for sharding prop. Why we need this? operators can generate multiple placement strategies, i.e. for matmul we have at least 4 possible shardings: `1. Shard(0), R 2. R, Shard(1) 3. Shard(1), Shard(0) 4. R, R`

We need to be able to choose from one of these options during runtime, and perform resharding with reasonable choices. This is why we are building a cost model for sharding here. In this PR we associate each possible sharding strategy with redistribute costs. For eager mode since we run ops eagerly, we simply perform a min cost selection. One can imagine if we have some global information the strategy selection would become more intelligient.


[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][8/N] Introduce cost model for sharding"


This PR adds some basic comm cost model for sharding prop. Why we need this? operators can generate multiple placement strategies, i.e. for matmul we have at least 4 possible shardings: `1. Shard(0), R 2. R, Shard(1) 3. Shard(1), Shard(0) 4. R, R`

We need to be able to choose from one of these options during runtime, and perform resharding with reasonable choices. This is why we are building a cost model for sharding here. In this PR we associate each possible sharding strategy with redistribute costs. For eager mode since we run ops eagerly, we simply perform a min cost selection. One can imagine if we have some global information the strategy selection would become more intelligient.


[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][8/N] Introduce cost model for sharding"


This PR adds some basic comm cost model for sharding prop. Why we need this? operators can generate multiple placement strategies, i.e. for matmul we have at least 4 possible shardings: `1. Shard(0), R 2. R, Shard(1) 3. Shard(1), Shard(0) 4. R, R`

We need to be able to choose from one of these options during runtime, and perform resharding with reasonable choices. This is why we are building a cost model for sharding here. In this PR we associate each possible sharding strategy with redistribute costs. For eager mode since we run ops eagerly, we simply perform a min cost selection. One can imagine if we have some global information the strategy selection would become more intelligient.


[ghstack-poisoned]

fduwjj commented on torch/distributed/_tensor/placement_types.py:
```diff
@@ -533,3 +533,9 @@ def from_dim_map(
 
         return cls(mesh, tuple(placements), tensor_meta=tensor_meta)
+
+    def is_replicated(self):
```
nice!

fduwjj commented on torch/distributed/_tensor/_collective_utils.py:
```diff
@@ -158,3 +160,128 @@ def mesh_all_to_all(
     return work
+
+
+def spec_to_bytes(spec: "placement_types.DTensorSpec") -> int:
```
To clarify, the result is in  bytes right?

wanchaol replied:
yep it's #num bytes

fduwjj commented on torch/distributed/_tensor/_collective_utils.py:
```diff
@@ -158,3 +160,128 @@ def mesh_all_to_all(
+
+def get_bandwidth_factor(mesh: DeviceMesh) -> List[float]:
+    # generate bandwidth factor for intra-host/inter-host communication pattern
+    factors = [1.0 * mesh.ndim]
```
Does this mean we are generating a list of one item or a list of mesh.ndim 1.0s?, IIUC, we want latter?

wanchaol replied:
good catch!!

fduwjj reviewed (commented):
Do we want to add at least one simple UT for the cost cost modeling?

wanchaol committed:
Update on "[dtensor][8/N] Introduce cost model for sharding"


This PR adds some basic comm cost model for sharding prop

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][8/N] Introduce cost model for sharding"


This PR adds some basic comm cost model for sharding prop

[ghstack-poisoned]

fduwjj commented on torch/distributed/_tensor/ops/tensor_ops.py:
```diff
@@ -121,6 +121,7 @@ def gen_bucketize_strategy(mesh: DeviceMesh, op_schema: OpSchema) -> StrategyTyp
     Just propagate input sharding, but expect replicated for boundaries input.
     """
     input_strategy = op_schema.args_schema[0]
+    boundaries_strategy = op_schema.args_schema[1]
```
Do we need this one?

fduwjj reviewed (approved):
LGTM

wanchaol committed:
Update on "[dtensor][8/N] Introduce cost model for sharding"


This PR adds some basic comm cost model for sharding prop

[ghstack-poisoned]
